shared/ssl.sh

```diff
@@ -0,0 +1,119 @@
+#
+# Fails under the following conditions:
+# - CA host cannot be reached
+# - CA already has signed certificate for this host
+# - SSL files already exist on this host
+#
+# Arguments:
+#   $1  (Optional) Certname to use, defaults to $HOSTNAME
+#
+# Optional environment variables:
+#   WAITFORCERT            Number of seconds to wait for certificate to be
+#                          signed, defaults to 120
+#   PUPPETSERVER_HOSTNAME  Hostname of Puppet Server CA, defaults to "puppet"
+#   SSLDIR                 Root directory to write files to, defaults to
+#                          "/etc/puppetlabs/puppet/ssl"
+
+msg() {
+    echo "($0) $1"
+}
+
+error() {
+    msg "Error: $1"
+    exit 1
+}
+
+### Verify dependencies available
+! type -p openssl > /dev/null && error "openssl not found on PATH"
+! type -p curl > /dev/null && error "curl not found on PATH"
+
+### Verify options are valid
+CERTNAME="${1:-${HOSTNAME}}"
+[ -z "${CERTNAME}" ] && error "certificate name must be non-empty value"
+PUPPETSERVER_HOSTNAME="${PUPPETSERVER_HOSTNAME:-puppet}"
+SSLDIR="${SSLDIR:-/etc/puppetlabs/puppet/ssl}"
+WAITFORCERT=${WAITFORCERT:-120}
+
+### Create directories and files
+PUBKEYDIR="${SSLDIR}/public_keys"
+PRIVKEYDIR="${SSLDIR}/private_keys"
+CSRDIR="${SSLDIR}/certificate_requests"
+CERTDIR="${SSLDIR}/certs"
+mkdir -p "${SSLDIR}" "${PUBKEYDIR}" "${PRIVKEYDIR}" "${CSRDIR}" "${CERTDIR}"
+PUBKEYFILE="${PUBKEYDIR}/${CERTNAME}.pem"
+PRIVKEYFILE="${PRIVKEYDIR}/${CERTNAME}.pem"
+CSRFILE="${CSRDIR}/${CERTNAME}.pem"
+CERTFILE="${CERTDIR}/${CERTNAME}.pem"
+CACERTFILE="${CERTDIR}/ca.pem"
+
+CA="https://${PUPPETSERVER_HOSTNAME}:8140/puppet-ca/v1"
+CERTSUBJECT="/CN=${CERTNAME}"
+CERTHEADER="-----BEGIN CERTIFICATE-----"
+
+### Print configuration for troubleshooting
+msg "Getting signed certificate for '${CERTNAME}' (${CERTSUBJECT})"
+msg "Using Puppet CA at '${CA}'"
+msg "Saving SSL files in '${SSLDIR}'"
+msg "Waiting up to ${WAITFORCERT} seconds for certificate to be signed"
+
+### Pre-checks before we begin:
+# 1) We can reach the CA host
+# 2) CA does not already have a signed certificate for this host
+output="$(curl --insecure --silent "${CA}/certificate/${CERTNAME}")"
+if [ $? -ne 0 ]; then
+    error "cannot reach CA host '${PUPPETSERVER_HOSTNAME}'"
+elif [ "$(echo "${output}" | head -1)" = "${CERTHEADER}" ]; then
+    error "CA already has signed certificate for '${CERTNAME}'"
+fi
+
+### Generate keys and CSR for this host
+[ -s "${PRIVKEYFILE}" ] && error "private key '${PRIVKEYFILE}' already exists"
+[ -s "${PUBKEYFILE}" ] && error "public key '${PUBKEYFILE}' already exists"
+[ -s "${CSRFILE}" ] && error "certificate request '${CSRFILE}' already exists"
+openssl genrsa -out "${PRIVKEYFILE}" 4096
+openssl rsa -in "${PRIVKEYFILE}" -pubout -out "${PUBKEYFILE}"
+openssl req -new -key "${PRIVKEYFILE}" -out "${CSRFILE}" -subj "${CERTSUBJECT}"
+
+### Submit CSR
+curl --insecure -X PUT -H "Content-Type: text/plain" \
```

Comment:
should we have `--silent` here to clean up output?

Reply:
I kind of went back and forth about it but decided to not silence it because otherwise it would hide useful troubleshooting messages.

And since this will only run one time, and be at the very beginning of the container, I figured it was better to have more info rather than less.

Same reason why I didn't quiet the openssl output.

Reply:
Fair. Mostly, I don't like the look of the progress bar in logs :P 

Looks like `--silent --show-error` would disable the progress bar but this is not a blocking issue.

Reply:
Haha I know, I want the output but not the bars.

Didn't know about that flag, I'll try it.

Reply:
That worked, no more noisy progress bars but still see errors: https://github.com/puppetlabs/pupperware/pull/33/commits/c02427fe1bc799e6a0183e14f2d56d970dbcb030